### PR TITLE
Feat wget ckpt

### PR DESCRIPTION
**Pull request by samsja**
live reco workig

### CODE REVIEW

File: src/zeroband/comms.py [hunk 1]
```diff
@@ -341,3 +350,55 @@ def get_global_pg(self, maybe_reinit: bool = False) -> dist.ProcessGroup:
...
+
+        self.enable = enable
+
+    def init_live_endpoint(self, store: dist.Store):
+        """
+        Put its own adress to the store so that other nodes can connect to it for live recovery
+        """
+        if not self.enable:
+            return
+        self.store = dist.PrefixStore("live_reco_adress", store)
```
Comment (JohannesHa) [inline src/zeroband/comms.py]: typo: address

Comment (JohannesHa) [reply src/zeroband/comms.py]: also everywhere else where "adress" is used haha

Comment (samsja) [reply src/zeroband/comms.py]: so I missed a d in address everywhere ? bravo sami for being consistently wrong

File: src/zeroband/diloco.py [hunk 2]
```diff
@@ -73,17 +73,21 @@ def _init_offloaded_optimizer(self, model):
...
+        self.elastic_device_mesh.maybe_reinit_global_pg(admit_joiners=False)
+        global_pg = self.elastic_device_mesh.global_pg
         for param_offloaded, param in zip(self.param_list_cpu, model.parameters()):
             if param.shape[0] == 0:
                 continue
-            param_offloaded.grad = param_offloaded.data - param.data.to(param_offloaded.device)
+            if fake:
+                param_offloaded.grad = param_offloaded.data - param.data.to(param_offloaded.device)
+            else:
+                param_offloaded.grad = torch.zeros_like(param_offloaded.data)
```
Comment (Jackmin801) [inline src/zeroband/diloco.py]: Eh isnt this reversed? If its fake we should zero otherwise we send diff?

File: src/zeroband/train.py [hunk 3]
```diff
@@ -86,8 +100,21 @@ class Config(BaseConfig):
...
     train: TrainConfig
 
-    ckpt: CkptConfig | None = None
-    resume: str | None = None
+    ckpt: CkptConfig = CkptConfig()
+
+    @model_validator(mode="after")
+    def live_reco_very_check(self):
+        if self.ckpt.live_recovery and self.diloco is None:
+            raise ValueError("Diloco must be set to use live recovery")
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Diloco must be set in order to use live recovery

I thought you meant that you must set live recovery when using diloco.

File: src/zeroband/comms.py [hunk 4]
```diff
@@ -346,3 +356,61 @@ def get_global_pg(self, maybe_reinit: bool = False) -> dist.ProcessGroup:
...
+
+        self._logger.debug(f"Live recovery adress: {LIVE_RECO_ADDR}:{self.port}")
+        self.store.set(f"adress_{self.world_info.global_unique_id}", f"{LIVE_RECO_ADDR}:{self.port}")
+
+    def get_adress(self, rank: int) -> str:
+        """Get the live recovery adress for a given rank."""
+        return self.store.get(f"adress_{rank}").decode("utf-8")
```
Comment (Jackmin801) [inline src/zeroband/comms.py]: Hrmm did u not refill on 'd's?

File: src/zeroband/comms.py [hunk 5]
```diff
@@ -20,6 +21,10 @@
     os.getenv("ZERO_BAND_EDM_HEARTBEAT_TIMEOUT_SECONDS", "10")
 )  # Time in seconds after which a node is considered dead if no heartbeat is received
 
+LIVE_RECO_PORT = os.environ.get("ZERO_BAND_LIVE_RECO_PORT", None)
```
Comment (Jackmin801) [inline src/zeroband/comms.py:24]: Hrmm should there be a default one?

File: src/zeroband/comms.py [hunk 6]
```diff
@@ -158,6 +165,7 @@ def _init_global_pg(self) -> None:
             self.world_info.global_world_size = int(self.global_store.get("world_size").decode("utf-8"))
             self.mesh_count = int(self.global_store.get("mesh_count").decode("utf-8"))
             prefix_store = dist.PrefixStore(f"mesh_{self.mesh_count}", self.global_store)
+            self.live_recovery.need_live_recovery = True if self.live_recovery.enable else False
```
Comment (Jackmin801) [inline src/zeroband/comms.py]: `self.live_recovery.enable` is already bool I think so this if else is redundant

File: src/zeroband/comms.py [hunk 7]
```diff
@@ -346,3 +356,61 @@ def get_global_pg(self, maybe_reinit: bool = False) -> dist.ProcessGroup:
...
+[ ] choose nodes to download to via bandwith
+[ ] avoid race condition when downloading while having another ckot
+[ ] If a node join but failed to donwload it will kill the whole training. Need to be fix
+
+
+"""
+
+
+class LiveRecovery:
+    def __init__(self, enable):
```
Comment (Jackmin801) [inline src/zeroband/comms.py]: typehint plox

File: src/zeroband/train.py [hunk 8]
```diff
@@ -189,18 +211,25 @@ def train(config: Config):
...
-        ckpt_manager.load(resume_ckpt_path=config.resume)
+        ckpt_manager.load(resume_ckpt_path=config.ckpt.resume)
 
-    model.train()
+    if elastic_device_mesh.live_recovery.need_live_recovery:
+        ckpt_manager.download_and_load_ckpt_from_peers(
+            elastic_device_mesh.live_recovery.get_adress(world_info.global_rank - 1)
+        )
+        elastic_device_mesh.live_recovery.need_live_recovery = False
+        diloco.fake_step(model)
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Do you need the load from resume_ckpt_path if you are doing live recovery? does it skip all the state live recovered state dict inits and only does dataloader loading?

File: src/zeroband/train.py [hunk 9]
```diff
@@ -223,9 +252,13 @@ def train(config: Config):
             logger.info(f"outer_step step: {training_progress.outer_step}")
 
         time_start_outer = time.perf_counter()
+
+        elastic_device_mesh.maybe_reinit_global_pg(admit_joiners=True)
+        # at the beginning of the outer steps we allow joiner to arrive.
```
Comment (Jackmin801) [inline src/zeroband/train.py]: at the beginning of the inner steps?

File: README.md [hunk 10]
```diff
@@ -109,3 +109,5 @@ HF_HUB_ETAG_TIMEOUT=500
 | `ZERO_BAND_GLOBAL_STORE_POLLING_INTERVAL_SECONDS` | Number of seconds between polls to the store when waiting for values | `0.1` |
 | `ZERO_BAND_EDM_HEARTBEAT_INTERVAL_SECONDS` | Interval in seconds between heartbeats | `2` |
 | `ZERO_BAND_EDM_HEARTBEAT_TIMEOUT_SECONDS` | Time in seconds after which a node is considered dead if no heartbeat is received | `10` |
+| `ZERO_BAND_LIVE_RECO_PORT` | Port number for the live recovery server | `8000` |
```
Comment (Jackmin801) [inline README.md]: oh theres a default here

File: src/zeroband/diloco.py (hunk 2 above)
Comment (samsja) [reply src/zeroband/diloco.py]: lol

File: src/zeroband/comms.py (hunk 5 above)
Comment (Jackmin801) [reply src/zeroband/comms.py:24]: set to 8000

**Comment by samsja** [comment]
working fine now Screenshot from 2024-10-08 12-40-21

**Review by Jackmin801 — APPROVED** [review]
NOICE! lgtm

**Pull request merged**